ksql-test-util/src/main/java/io/confluent/ksql/test/util/ImmutableTester.java

```diff
@@ -0,0 +1,173 @@
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.test.util;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.Immutable;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalDouble;
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.function.Predicate;
+
+public final class ImmutableTester {
+
+  private static final List<Predicate<Class<?>>> STD_IMMUTABLE_TYPES = ImmutableList
+      .<Predicate<Class<?>>>builder()
+      .add(Class::isPrimitive)
+      .add(Class::isEnum)
+      .add(String.class::isAssignableFrom)
+      .add(Optional.class::isAssignableFrom)
+      .add(OptionalInt.class::isAssignableFrom)
+      .add(OptionalLong.class::isAssignableFrom)
+      .add(OptionalDouble.class::isAssignableFrom)
+      .build();
+
+  private final List<Predicate<Class<?>>> knownImmutables = new ArrayList<>(STD_IMMUTABLE_TYPES);
+
+  public ImmutableTester withKnownImmutableTypes(final Set<Class<?>> immutable) {
+    immutable.forEach(type -> knownImmutables.add(t -> t.equals(type)));
+    return this;
+  }
+
+  public ImmutableTester withKnownImmutableTypes(
+      final Collection<Predicate<Class<?>>> immutable
+  ) {
+    knownImmutables.addAll(immutable);
+    return this;
+  }
+
+  public ImmutableTester withKnownImmutableType(final Class<?> immutable) {
+    knownImmutables.add(t -> t.equals(immutable));
+    return this;
+  }
+
+  public ImmutableTester withKnownImmutableType(
+      final Predicate<Class<?>> immutable
+  ) {
+    knownImmutables.add(immutable);
+    return this;
+  }
+
+  public void test(final Class<?> typeUnderTest) {
+    final String className = typeUnderTest.getSimpleName();
+
+    final Predicate<Class<?>> knownImmutable = knownImmutables
+        .stream()
+        .reduce(type -> false, Predicate::or);
+
+    assertThat("@Immutable annotation missing",
+        typeUnderTest.isAnnotationPresent(Immutable.class),
+        is(true));
+
+    Arrays.stream(typeUnderTest.getDeclaredFields())
+        .filter(excludeJaCoCoInjectedFields())
+        .forEach(field -> {
+          assertThat(className + ": field not final: " + field.getName(),
+              Modifier.isFinal(field.getModifiers()),
+              is(true));
+
+          try {
```

Comment:
Why do we need this try-catch around `checkImmutableType()`? (What error are we trying to catch?)

Reply:
It's catching a 'this is not an immutable type' assertion and ensuring the thrown exception includes the field name.